File: site/src/data/formato.ts
/**
 * Numbers and dates as they are written in Brazil, for the values the components render.
 *
 * The generated pages get the same treatment from the pipeline; this is the other half, so
 * a length in a card and the same length in a table read identically.
 */

/**
 * Line lengths are shown to one decimal. The sources place the same station up to a few
 * hundred metres apart, so writing a length to the metre would claim a precision the data
 * does not have — the full value stays in `network.json`.
 */
const CASAS_EXTENSAO = 1;

export const numero = (valor: number, casas = 0): string =>
  new Intl.NumberFormat("pt-BR", {
    minimumFractionDigits: casas,
    maximumFractionDigits: casas,
  }).format(valor);

export const quilometros = (valor: number): string => `${numero(valor, CASAS_EXTENSAO)} km`;

/** `1974-09-14` → `14/09/1974`. Parsed as a plain date, never shifted by a time zone. */
export const data = (iso: string): string => {
  const partes = /^(\d{4})-(\d{2})-(\d{2})/.exec(iso);
  return partes ? `${partes[3]}/${partes[2]}/${partes[1]}` : iso;
};

File: pipeline/src/transporte_sp/export/formato.py
"""Numbers and dates as they are written in Brazil.

Comma for the decimal, dot for the thousands, ``dd/mm/aaaa`` for dates. Done here rather
than with ``locale`` because that would make the generated pages depend on which locales
happen to be installed on the machine running the pipeline.
"""

from __future__ import annotations

from datetime import date

# Line lengths are published to one decimal. The sources place the same station up to a few
# hundred metres apart, so a length written to the metre would claim a precision the data
# does not have. The full value stays in network.json.
CASAS_EXTENSAO = 1


def numero(valor: float, casas: int = 0) -> str:
    """``1234.5`` → ``1.234,5``."""
    texto = f"{valor:,.{casas}f}"
    return texto.replace(",", "\x00").replace(".", ",").replace("\x00", ".")


def quilometros(valor: float) -> str:
    """``12.765`` → ``12,8 km``."""
    return f"{numero(valor, CASAS_EXTENSAO)} km"


def data(valor: date | str) -> str:
    """``1974-09-14`` → ``14/09/1974``."""
    if isinstance(valor, str):
        try:
            valor = date.fromisoformat(valor[:10])
        except ValueError:
            return valor
    return valor.strftime("%d/%m/%Y")
